Main branch

AlexBossov opened the pull request.

AlexBossov committed:
В класс DeadLine добвлены сеттеры

AlexBossov committed:
1) Изменено название класса GroupOfTasks на TaskGroup
2) В класс TaskGroup добавлена функция ToString, удалены функции Print и PrintToFile
3) Изменены все функции TaskGroup(Переписаны с помощью LINQ + изменены исключения)
4) Так же переписана функция ToString в классе Task(StingBuilder)
5) Теперь отлов эксепшенов происходит только в Main

AlexBossov commented:
Кстати, после того как у класса TaskGroup есть удобная функция ToString чисто-теоретически могу попробовать сделать вот эту часть задания
Группировка задач - возможность создавать группы задач
     /create-group group-name - создает группу для задач
     /delete-group group-name - удаляет группу с заданным именем
     /add-to-group id group-name - добавляет таску с указанным id в группу с указанным именем
     /delete-from-group id group-name - удаляет задачу c группы
  Задачи, которые находятся в группе, должны при выполнении /all отображаться вложенным списком
    /completed group-name - выводит все выполненные в группе задачи
Есть ли в этом смысл?

ronimizy commented:
Конечно! Практика никогда лишней не бывает)

ronimizy commented on ToDoList/Program.cs:
```diff
@@ -1,93 +1,96 @@
                     {
                         case "/all":
-                            groupOfTasks.Print();
+                            Console.Write(taskGroup.ToString());
```
можно не вызывать .ToString() явно, просто передать taskGroup, он сам вызовется в дальнейшем внутри Console.Write

ronimizy commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,171 @@
+            foreach (var task in _tasks)
+            {
+                stringBuilder.Append($". {task.ToString()}\n");
+                if (task.SubTasks.Count.Equals(0)) continue;
```
эта проверка не нужна)

ronimizy commented on ToDoList/Task.cs:
```diff
@@ -44,23 +45,30 @@ public void AddSubTask(Task task)
+            stringBuilder.Append($"(id = {Id})");
 
-            task += $"(id = {Id})";
+            var task = stringBuilder.ToString();
             return task;
```
стоит сразу возвращать 
```cs
stringBuilder.ToString()
```

ronimizy commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,171 @@
+                    stringBuilder.Append($"\t* {subTask.ToString()}\n");
+            }
+
+            var tmp = stringBuilder.ToString();
+            return tmp;
```
так же про возврат

ronimizy commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,171 @@
+            var task = _tasks.FirstOrDefault(t => t.Id == id || t.SubTasks.Any(s => s.Id == id));
+
+            if (task is null)
+                throw new ArgumentOutOfRangeException(nameof(id));
```
Какой-то не информативный эксепшн)

ronimizy commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,171 @@
+            var task = _tasks.FirstOrDefault(t => t.Id == id || t.SubTasks.Any(s => s.Id == id));
+
+            if (task is null)
+                throw new ArgumentOutOfRangeException(nameof(id));
```
то же про информативность

ronimizy commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,171 @@
+            var task = _tasks.FirstOrDefault(t => t.Id == id);
+
+            if (task is null)
+                throw new ArgumentOutOfRangeException(nameof(id));
```
то же про информативность

ronimizy reviewed (changes requested):
Стоит сделать постоянство с LINQ, либо методы, либо синтаксический (советую методы)
Методы:
```cs
task.SubTasks.Where(subTask => subTask.Id.Equals(id)
```

Синтаксис:
```cs
from subTask in task.SubTasks
                     where subTask.Deadline.Day == DateTime.Now.Day &&
                           subTask.Deadline.Month == DateTime.Now.Month &&
                           subTask.Deadline.Year == (ulong) DateTime.Now.Year
                     select subTask;
```

Так же заметил, что у тебя часто повторяется парсинг, попробуй вынести код 
```cs
var d = Convert.ToUInt16(line.Substring(found - 2, 2));
var m = Convert.ToUInt16(line.Substring(found + 1, 2));
var y = Convert.ToUInt64(line.Substring(found + 4, 4));
```

AlexBossov commented:
> Так же заметил, что у тебя часто повторяется парсинг, попробуй вынести код 
> var d = Convert.ToUInt16(line.Substring(found - 2, 2));
> var m = Convert.ToUInt16(line.Substring(found + 1, 2));
> var y = Convert.ToUInt64(line.Substring(found + 4, 4));

Будет ли хороошим решением сделать оператор преобразования типов в калссе DeadLine из string в DeadLine?)
public static implicit operator DeadLine(string s)
        {
           return new DeadLine(Convert.ToUInt16(s[..2]),
               Convert.ToUInt16(s.Substring(3, 2)), 
                   Convert.ToUInt64(s.Substring(6, 4)));
        }
Или лучше сделать метод FromString() в классе DeadLine?

ronimizy commented:
Я бы просто сделал конструктор из строки)

AlexBossov commented:
> Я бы просто сделал конструктор из строки)

Лол, логично, почему мне это не пришло в голову)))

AlexBossov commented:
> Конечно! Практика никогда лишней не бывает)

А стоит ли делать отдельный класс для этого или сделать List из TaskGroup? ))))))
И еще может можно как-то все это проще реализовать?)

ronimizy commented:
Подумай как реализация будет чище, чтобы одни и те же функции не повторялись

AlexBossov committed:
1) LINQ теперь реалозовано одним способом
2) В класс DeadLine добавлен конструктор с параметром в виде string
3) Испавлены выбросы исключений

ronimizy commented on ToDoList/Program.cs:
```diff
@@ -23,7 +27,8 @@ private static void Main(string[] args)
                     {
                         case "/all":
-                            Console.Write(taskGroup.ToString());
+                            Console.Write(taskGroup
+                                .ToString()); // Если делаю без ToString(), то выводится ToDoList.TaskGroup
```
В классе TaskGroup, и в остальных местах, функцию .ToString() нужно явно переопределить кейвордом ```override```

ronimizy reviewed (changes requested):
Советую посмотреть [видео](https://www.youtube.com/watch?v=aSJD_PcWPjU&feature=youtu.be) от Фреди про гх, как минимум про часть с неймингом коммитов

AlexBossov commented:
> Конечно! Практика никогда лишней не бывает)

Вопрос на счет групп, если задача есть в группе, она выводится два раза(т.е. без группы, потом с группой) или же 1 раз(т.е. только в определенной группе)?

ronimizy commented:
Я делал чтобы выводилась один раз, в принципе особо смысла чтобы выводилась два раза нет

AlexBossov commented:
> Я делал чтобы выводилась один раз, в принципе особо смысла чтобы выводилась два раза нет

Хорошо, а /del удалять может из группы в том числе? И /delete=from-group удаляет задачу из группы и кидает в обычный список?

ronimizy commented:
Верно)

AlexBossov committed:
feat: grouping
fix: functions of TaskGroup

FrediKats commented on ToDoList/DeadLine.cs:
```diff
@@ -1,17 +1,36 @@
         public DeadLine(ushort day = 0, ushort month = 0, ulong year = 0)
         {
-            if (day > 31 || month > 12 || year == 0 && year < 2021) return;
+            if (day > 31 || month > 12 || year is 0 and < 2021) return;
```
В таких случаях ошибку стоит бросать.

AlexBossov replied:
да, изначально бросал)

FrediKats commented on ToDoList/DeadLine.cs:
```diff
@@ -1,17 +1,36 @@
+            Year = year;
+        }
+
+        public DeadLine(string deadLine)
```
Кажется, ты изобрел DateTime и DateTime.Parse

AlexBossov replied:
К сожалению, понял это уже под конец написания кода)

AlexBossov committed:
fix: command "/save"

FrediKats commented on ToDoList/Program.cs:
```diff
@@ -1,108 +1,188 @@
                     {
                         case "/all":
-                            groupOfTasks.Print();
+                            Console.Write(taskGroups.First().ToString()); // First - задачи без группы
```
Если ты понимаешь, что нужно такой коммент оставить - значит нужно что-то изменить в коде)
Например, добавить метод, который вернет задачи без групп. И нормальное название давать. Комменты - это звоночек о проблемах. Иногда это оправдано. Тут - нет.

FrediKats commented on ToDoList/Program.cs:
```diff
@@ -1,108 +1,188 @@
                         case "/all":
-                            groupOfTasks.Print();
+                            Console.Write(taskGroups.First().ToString()); // First - задачи без группы
+                            for (index = 1; index < taskGroups.Count; index++)
+                            {
+                                var taskGroup = taskGroups[index];
+
+                                Console.Write($"{taskGroup.Name}\n   ");
+                                var tmp = new StringBuilder(taskGroup.ToString());
+                                tmp.Replace("\n", "\n   ");
+                                Console.WriteLine(tmp);
+                            }
+
```
Было бы намного проще читать, если бы код в case'ах был вынесен в методы.

FrediKats commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,186 @@
+
+        public void Add(Task task) // /add task-info
+        {
+            if (_tasks.Any(task1 => task.Text.Equals(task1.Text))) return;
```
Опять же, если ты хочешь, чтобы не добавлялись одинаковые - бросай ошибки. Иначе я бы очень удивился если бы задач было 5, я добавил одну и .... их 5)

FrediKats commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,186 @@
+        {
+            if (_tasks.Any(task1 => task.Text.Equals(task1.Text))) return;
+            _tasks.Add(task);
+            Task.IncrementIdentifier();
```
Думаю, это должно происходить автоматически, иначе есть вероятность, что ты забудешь это сделать.

FrediKats commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,186 @@
+                    return false;
+
+                subTask.Completed = true;
+                task.CountOfCompletedSubTasks++;
```
Это можно сделать вычисляемым полем, иначе опять же можешь забыть его обновлять.
CountOfCompletedSubTasks => SubTasks.Count(st => st.Completed)

FrediKats commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,186 @@
+
+                subTask.Completed = true;
+                task.CountOfCompletedSubTasks++;
+                if (task.CountOfCompletedSubTasks.Equals((ushort) task.SubTasks.Count) &&
+                    !task.SubTasks.Count.Equals(0)) task.Completed = true;
```
Вот такие штуки похожи на логику таски. Я бы сделал у таски метод MarkCompleted, чтобы сама таска уже решала что она там внутри хочет обновить и засетить.

+ ты решил числа сравнить через Equals?) Это уже оверинжениринг. Лучше выразить это через LINQ task.SubTasks.All(st => st.Completed)

FrediKats commented on ToDoList/TaskGroup.cs:
```diff
@@ -0,0 +1,186 @@
+
+        public string Today() // /today
+        {
+            var todayTasks = _tasks.Where(task => task.Deadline.Day == DateTime.Now.Day &&
+                                                  task.Deadline.Month == DateTime.Now.Month &&
+                                                  task.Deadline.Year == (ulong) DateTime.Now.Year);
```
А почему ты не использовать DateTime? У него есть свойство .Date, которое вернет дату без времени и можно сравнивать.
